From e8fc7399a52e022813ec33ec6db77ced3538bf34 Mon Sep 17 00:00:00 2001
From: Emilien <4016501+unixfox@users.noreply.github.com>
Date: Sat, 26 Oct 2024 20:53:28 +0200
Subject: [PATCH 1/1] range header to query string

---
 assets/js/player.js | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/assets/js/player.js b/assets/js/player.js
index 353a5296..18a870d7 100644
--- a/assets/js/player.js
+++ b/assets/js/player.js
@@ -42,6 +42,22 @@ embed_url = location.origin + '/embed/' + video_data.id + embed_url.search;
 var save_player_pos_key = 'save_player_pos';
 
 videojs.Vhs.xhr.beforeRequest = function(options) {
+    // send range in the query string and with POST method
+    // taken from Freetube:
+    // https://github.com/FreeTubeApp/FreeTube/blob/development/src/renderer/components/ft-shaka-video-player/ft-shaka-video-player.js#L1096
+    if (options.uri.includes("videoplayback")) {
+        options.method = 'POST'
+        options.body = new Uint8Array([0x78, 0]) // protobuf: { 15: 0 } (no idea what it means but this is what YouTube uses)
+        if (options.headers) {
+            if (options.headers.Range) {
+                options.uri += `&range=${options.headers.Range.split('=')[1]}`
+                delete options.headers.range
+            }
+        }
+
+        options.uri += '&alr=yes'
+    }
+
     // set local if requested not videoplayback
     if (!options.uri.includes('videoplayback')) {
         if (!options.uri.includes('local=true'))
-- 
2.47.0

