From 943effeb4dc0235ae825db3df59843771f3f2106 Mon Sep 17 00:00:00 2001
From: Emilien Devos <4016501+unixfox@users.noreply.github.com>
Date: Tue, 15 Aug 2023 00:00:45 +0200
Subject: [PATCH 1/1] message registration disabled

---
 src/invidious/routes/login.cr | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/invidious/routes/login.cr b/src/invidious/routes/login.cr
index add9f75d..9fb21dfa 100644
--- a/src/invidious/routes/login.cr
+++ b/src/invidious/routes/login.cr
@@ -73,7 +73,7 @@ module Invidious::Routes::Login
         end
       else
         if !CONFIG.registration_enabled
-          return error_template(400, "Registration has been disabled by administrator.")
+          return error_template(400, "Please read why you can't use an account anymore: https://gist.github.com/yewtudotbe/c16a69ddad88a37c2a364a5ff5359197")
         end
 
         if password.empty?
-- 
2.41.0

