From a0c481387a701c85b0a1b70af3d642e3fe375e1f Mon Sep 17 00:00:00 2001
From: Emilien Devos <contact@emiliendevos.be>
Date: Fri, 10 Jun 2022 17:38:49 +0200
Subject: [PATCH 1/1] decrease buffer seconds for saving bandwidth

---
 assets/js/player.js | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/assets/js/player.js b/assets/js/player.js
index 7d099e66..02c8ae26 100644
--- a/assets/js/player.js
+++ b/assets/js/player.js
@@ -50,6 +50,9 @@ videojs.Vhs.xhr.beforeRequest = function(options) {
     return options;
 };
 
+videojs.Vhs.GOAL_BUFFER_LENGTH = 20;
+videojs.Vhs.MAX_GOAL_BUFFER_LENGTH = 30;
+
 var player = videojs('player', options);
 
 player.on('error', function () {
-- 
2.36.1

