From 8518074d993f1cedc5b0c51852ad6c53e7c39e18 Mon Sep 17 00:00:00 2001
From: Emilien Devos <contact@emiliendevos.be>
Date: Tue, 8 Mar 2022 00:17:28 +0100
Subject: [PATCH 1/1] decrease buffer seconds for saving bandwidth

---
 assets/js/player.js | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/assets/js/player.js b/assets/js/player.js
index a1a2cd16..489205cf 100644
--- a/assets/js/player.js
+++ b/assets/js/player.js
@@ -47,6 +47,9 @@ videojs.Vhs.xhr.beforeRequest = function(options) {
     return options;
 };
 
+videojs.Vhs.GOAL_BUFFER_LENGTH = 20;
+videojs.Vhs.MAX_GOAL_BUFFER_LENGTH = 30;
+
 var player = videojs('player', options);
 
 /**
-- 
2.35.1

