From 4d8a64495fda01898a13b2942105ebc46b6c840f Mon Sep 17 00:00:00 2001
From: Emilien Devos <contact@emiliendevos.be>
Date: Tue, 25 Jan 2022 20:55:48 +0100
Subject: [PATCH 1/1] disable watch history for now

---
 src/invidious/database/users.cr | 2 --
 1 file changed, 2 deletions(-)

diff --git a/src/invidious/database/users.cr b/src/invidious/database/users.cr
index 53724dbf..7cb36040 100644
--- a/src/invidious/database/users.cr
+++ b/src/invidious/database/users.cr
@@ -56,7 +56,6 @@ module Invidious::Database::Users
       WHERE email = $2
     SQL
 
-    PG_DB.exec(request, vid, user.email)
   end
 
   def mark_unwatched(user : User, vid : String)
@@ -66,7 +65,6 @@ module Invidious::Database::Users
       WHERE email = $2
     SQL
 
-    PG_DB.exec(request, vid, user.email)
   end
 
   def clear_watch_history(user : User)
-- 
2.34.1

