From 09e382465cebd61aa7a8b67b0a87a939e8ab5a03 Mon Sep 17 00:00:00 2001
From: Emilien Devos <contact@emiliendevos.be>
Date: Wed, 26 Jan 2022 21:47:00 +0100
Subject: [PATCH 1/1] disable notifications

---
 src/invidious/database/users.cr | 2 --
 1 file changed, 2 deletions(-)

diff --git a/src/invidious/database/users.cr b/src/invidious/database/users.cr
index 53724dbf..3752a213 100644
--- a/src/invidious/database/users.cr
+++ b/src/invidious/database/users.cr
@@ -127,7 +127,6 @@ module Invidious::Database::Users
       WHERE $2 = ANY(subscriptions)
     SQL
 
-    PG_DB.exec(request, video.id, video.ucid)
   end
 
   def remove_notification(user : User, vid : String)
@@ -137,7 +136,6 @@ module Invidious::Database::Users
       WHERE email = $2
     SQL
 
-    PG_DB.exec(request, vid, user.email)
   end
 
   def clear_notifications(user : User)
-- 
2.34.1

