From 3ad7b342cb24b16f9885b18ee5321cd695ac7dcd Mon Sep 17 00:00:00 2001
From: Emilien Devos <contact@emiliendevos.be>
Date: Wed, 23 Feb 2022 23:48:17 +0100
Subject: [PATCH 1/1] reuse port by default

---
 src/invidious.cr | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/invidious.cr b/src/invidious.cr
index d4878759..5a2cdce0 100644
--- a/src/invidious.cr
+++ b/src/invidious.cr
@@ -562,4 +562,8 @@ Kemal.config.app_name = "Invidious"
   Kemal.config.env = "production" if !ENV.has_key?("KEMAL_ENV")
 {% end %}
 
-Kemal.run
+Kemal.run do |config|
+  server = config.server.not_nil!
+  server.bind_tcp Kemal.config.host_binding, Kemal.config.port, reuse_port: true
+  server.bind_tcp Kemal.config.host_binding, 0
+end
\ No newline at end of file
-- 
2.35.1

